src/Paprika/Data/SlottedArray.cs

```diff
@@ -906,78 +910,198 @@ private void MarkAsDeleted(int index)
         {
-            var slot = GetSlotRef(i);
-            if (slot.IsDeleted)
-            {
-                writeAt = i; // write at this slot
-                readTo = slot.ItemAddress; // mark it as read already
-                writtenTo = (ushort)(i == 0 ? _data.Length : GetSlotRef(i - 1).ItemAddress); // memoize the previous
-
-                // move to next
-                i++;
-                break;
-            }
-
-            alive++;
+            VectorizedDefragment256();
         }
-
-        for (; i < count; i++)
+        else if (Vector128.IsHardwareAccelerated)
         {
-            var slot = GetSlotRef(i);
-            var addr = slot.ItemAddress;
-
-            if (!slot.IsDeleted)
-            {
-                alive++;
-
-                var length = readTo - addr;
-
-                if (length > 0)
-                {
-                    var source = _data.Slice(addr, length);
-                    writtenTo = (ushort)(writtenTo - length);
-                    var destination = _data.Slice(writtenTo, length);
-                    source.CopyTo(destination);
-                }
-
-                // Copy hash
-                GetHashRef(writeAt) = GetHashRef(i);
-
-                // Copy everything, just overwrite the address
-                ref var destinationSlot = ref GetSlotRef(writeAt);
-                destinationSlot.KeyPreamble = slot.KeyPreamble;
-                destinationSlot.ItemAddress = writtenTo;
-
-                writeAt++;
-            }
-
-            // Memoize to what is read to
-            readTo = addr;
+            VectorizedDefragment128();
         }
+        else
+        {
+            ThrowNoVectorSupport();
+        }
+    }
 
-        // Finalize by setting the header
-        _header.Low = (ushort)(alive * Slot.TotalSize);
-        _header.High = (ushort)(_data.Length - writtenTo);
-        _header.Deleted = 0;
+     private void VectorizedDefragment256()
+     {
+         var count = _header.Low / Slot.TotalSize;
+         var jump = DoubleVectorSize / sizeof(ushort);
+         var aligned = AlignToDoubleVectorSize(_header.Low) / sizeof(ushort);
+
+         // Start from the first slot vector
+         var writeTo = HashesPerVector;
+
+         var preambleMask = Slot.GetKeyPreambleMaskAsVector256();
+         var deletedMask = Slot.GetKeyPreambleDeleteAsVector256();
+
+         ref var d = ref Unsafe.As<byte, ushort>(ref MemoryMarshal.GetReference(_data));
+         var i = writeTo;
+
+         while (i < aligned)
+         {
+             if (i + jump < aligned)
+             {
+                 var slotData = Vector256.LoadUnsafe(ref d, (UIntPtr)i);
+
+                 // Ignore all the even entries from the vector as the preamble data is contained within
+                 // the first 2 bytes (ushort) out of the 4 bytes slot.
+                 slotData = Avx2.BlendVariable(slotData, Vector256<ushort>.Zero, Vector256.Create((ushort)0b_1010_1010_1010_1010));
```

Comment:
Why is this needed? What does it do? Isn't it a bug? There's `BitwiseAnd` underneath so it should remove what is not needed? The comment states that the slot is 4 bytes long, but after #376 this is no longer the case in a sense that the preambles and hashes are grouped into vector sizes, see: https://github.com/NethermindEth/Paprika/blob/main/docs/design.md#slottedarray-layout

Reply:
Ah, I earlier assumed it to be 2 bytes and recently added this mask. I think I got misled by the total size parameter within Slot:

`public const int TotalSize = Size + sizeof(ushort);`

Just realized that it mentions in the comment that it's total size of hash and slot together, missed that completely. Maybe I will rename the parameter to `TotalSizeWithHash` for better clarity? From the current name TotalSize and Size look exactly similar :sweat_smile: